### PR TITLE
Libssh2 eating error codes

### PR DESCRIPTION
**Pull request by kevinbackhouse**

### CODE REVIEW

File: ql_demos/cpp/libssh2_eating_error_codes/README.md [hunk 1]
```diff
@@ -0,0 +1,9 @@
+# Eating error codes in libssh2
+
+Use this [Snapshot](https://downloads.lgtm.com/snapshots/cpp/libssh2/libssh2_libssh2_C_C++_38bf7ce.zip) for the demo.
+
+This demo shows how to develop, step-by-step, the query from the [blog post](https://blog.semmle.com/libssh2-integer-overflow/) about libssh2 CVE-2019-13115. This query did not find the bug that caused the CVE. It is instead about doing variant analysis on a bug which we noticed on the development branch of libssh2. We sent the query results to the libssh2 development team and they were able to fix all the variants before the next version of libssh2 was released.
```
Comment (jf205) [inline ql_demos/cpp/libssh2_eating_error_codes/README.md]:
```suggestion
This demo shows how to develop, step-by-step, the query from the [blog post](https://blog.semmle.com/libssh2-integer-overflow/) about libssh2 CVE-2019-13115. This query did not find the bug that caused the CVE. It is instead about doing variant analysis on a bug that we noticed on the development branch of libssh2. We sent the query results to the libssh2 development team and they were able to fix all the variants before the next version of libssh2 was released.
```

File: ql_demos/cpp/libssh2_eating_error_codes/README.md [hunk 2]
```diff
@@ -0,0 +1,9 @@
...
+
+[This](https://lgtm.com/projects/g/libssh2/libssh2/snapshot/6e2f5563c80521b3cde72a6fcdb675c2e085f9cf/files/src/hostkey.c?sort=name&dir=ASC&mode=heatmap&__hstc=70225743.5fa8704c8874c6eafaef219923a26734.1534954774206.1564532078978.1564925733575.72&__hssc=70225743.2.1565139962633&__hsfp=997709570#L677) is an example of the bug. The problem is that `_libssh2_get_c_string` returns an negative integer as an error code, but the type of `r_len` is `unsigned int`, so the error code is accidentally ignored.
```
Comment (jf205) [inline ql_demos/cpp/libssh2_eating_error_codes/README.md]:
```suggestion
[This](https://lgtm.com/projects/g/libssh2/libssh2/snapshot/6e2f5563c80521b3cde72a6fcdb675c2e085f9cf/files/src/hostkey.c?sort=name&dir=ASC&mode=heatmap&__hstc=70225743.5fa8704c8874c6eafaef219923a26734.1534954774206.1564532078978.1564925733575.72&__hssc=70225743.2.1565139962633&__hsfp=997709570#L677) is an example of the bug. The problem is that `_libssh2_get_c_string` returns a negative integer as an error code, but the type of `r_len` is `unsigned int`, so the error code is accidentally ignored.
```

File: ql_demos/cpp/libssh2_eating_error_codes/README.md [hunk 3]
```diff
@@ -0,0 +1,9 @@
...
+
+For a shorter demo, stop at step 02. Steps 03 and 04 make the query more sophisticated by adding local dataflow and range analysis.
```
Comment (jf205) [inline ql_demos/cpp/libssh2_eating_error_codes/README.md]:
```suggestion
For a shorter demo, stop at step 02. Steps 03 and 04 make the query more sophisticated by adding local data flow and range analysis.
```

File: ql_demos/cpp/libssh2_eating_error_codes/README.md [hunk 4]
```diff
@@ -0,0 +1,9 @@
+# Eating error codes in libssh2
+
+Use this [Snapshot](https://downloads.lgtm.com/snapshots/cpp/libssh2/libssh2_libssh2_C_C++_38bf7ce.zip) for the demo.
```
Comment (jf205) [inline ql_demos/cpp/libssh2_eating_error_codes/README.md]:
```suggestion
Download this [snapshot](https://downloads.lgtm.com/snapshots/cpp/libssh2/libssh2_libssh2_C_C++_38bf7ce.zip) for the demo.
```

File: ql_demos/cpp/libssh2_eating_error_codes/README.md (hunk 1 above)
Comment (kevinbackhouse) [reply ql_demos/cpp/libssh2_eating_error_codes/README.md]: I am happy to accept this change, but please could you explain why "that" is better than "which" here? ("which vs that" is something that I am a bit fuzzy on.)

Comment (jf205) [reply ql_demos/cpp/libssh2_eating_error_codes/README.md]: I follow the rationale outlined here: https://www.writersdigest.com/online-editor/which-vs-that

As I understand it, if you are using 'which' you should be able to drop the clause without changing the meaning of the sentence. That doesn't seem to be the case here, as we need the clause to identify the bug specifically as the one on the development branch. Without the clause the sentence doesn't make much sense.

[As a rough rule, 'which' would normally follow a comma too.]

It's one of those things that probably doesn't matter too much, but I've got into the habit of spotting and correcting. I'm sure you could get away without making the change.

Hope that helps?

**Comment by kevinbackhouse** [comment]
@jf205: Thanks for reviewing! I had forgotten about this PR. I have merge capability on this repo, but I have forgotten who else has. I don't have admin privileges though. We should definitely enable you to merge PRs too.

File: ql_demos/cpp/libssh2_eating_error_codes/README.md (hunk 1 above)
Comment (kevinbackhouse) [reply ql_demos/cpp/libssh2_eating_error_codes/README.md]: Thanks!

**Comment by jf205** [comment]
No worries. We should also get #9 merged too, if you are happy that the queries don't require a technical review.

**Pull request merged**